pip/pip-281.md

```diff
@@ -0,0 +1,103 @@
+# Title: [io] Add notifyError method on PushSource
+
+## Motivation
+
+In function framework, when [source.read()](https://github.com/apache/pulsar/blob/f7c0b3c49c9ad8c28d0b00aa30d727850eb8bc04/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L496-L506) method throw exception, it will trigger close function instance. If it is in the k8s environment, it will be restarted
+
+On io source connector, we provide [PushSource](https://github.com/apache/pulsar/blob/branch-3.0/pulsar-io/core/src/main/java/org/apache/pulsar/io/core/PushSource.java) class for users to use, and users can extend this class to quickly implement the push message model.
+It overrides the `read` method and provides the `consume` method for the user to call.
+
+However, if the source connector that extends from the class, 
+it cannot notify the function framework if it encounters an exception while consuming data internally, 
+in other words, the function call `source.read()` never triggers an exception and never exits the process.
+
+
+## Goals
+
+Add `notifyError` method on PushSource, This method can receive an exception and put the exception in the queue. The next time an exception is `read`, will throws it.
```

Comment:
```suggestion
Add the `notifyError` method on PushSource. This method can receive an exception and put the exception in the queue. The next time an exception is `read`, it will throw it.
```

throw "it", throw what?